File: src/enginedemo/ios/BalanceAndPanDemo.m
/*
 Copyright (c) 2010-2011 Per Gantelius
 
 This software is provided 'as-is', without any express or implied
 warranty. In no event will the authors be held liable for any damages
 arising from the use of this software.
 
 Permission is granted to anyone to use this software for any purpose,
 including commercial applications, and to alter it and redistribute it
 freely, subject to the following restrictions:
 
 1. The origin of this software must not be misrepresented; you must not
 claim that you wrote the original software. If you use this software
 in a product, an acknowledgment in the product documentation would be
 appreciated but is not required.
 
 2. Altered source versions must be plainly marked as such, and must not be
 misrepresented as being the original software.
 
 3. This notice may not be removed or altered from any source
 distribution.
 */

#import "BalanceAndPanDemo.h"

@implementation BalanceAndPanDemo

@synthesize panSlider;
@synthesize m_meterBarLeft;
@synthesize m_meterBarRight;

-(NSString*)getName
{
    return @"Balance and pan";
}

-(NSString*)getDescription
{
    return @"";
}

-(NSString*)getInstructions
{
    return @"";
}

-(void)initialize
{
    panSlider.value = 0.0f;
    m_waveBankHandle = kwlWaveBankLoad([DemoBase getResourcePath:@"sfx.kwb"]);
    m_eventHandle = kwlEventGetHandle("balancedemo/mono_and_stereo");
    kwlEventStartFade(m_eventHandle, 1.0f);
    kwlEventSetBalance(m_eventHandle, 0.0f);
    kwlLevelMeteringSetEnabled(1);
}

-(void)deinitialize
{
    kwlEventStop(m_eventHandle);
    kwlEventRelease(m_eventHandle);
    kwlWaveBankUnload(m_waveBankHandle);
    m_eventHandle = KWL_INVALID_HANDLE;
    m_waveBankHandle = KWL_INVALID_HANDLE;
    kwlLevelMeteringSetEnabled(0);
}

-(void)update:(float)timeStep
{
    m_meterBarLeft.progress = kwlGetLevelLeft();
    m_meterBarRight.progress = kwlGetLevelRight();
}

- (IBAction)panChanged:(id)sender
{
    UISlider* s = (UISlider*)sender;
    kwlEventSetBalance(m_eventHandle, s.value);
}


@end

File: src/enginedemo/sdl/BalanceAndPanDemo.cpp
#include "BalanceAndPanDemo.h"
#include "KowalskiDemoManager.h"
#include "assert.h"
#include "fileutil.h"

BalanceAndPanDemo::BalanceAndPanDemo() :
    m_eventHandle(KWL_INVALID_HANDLE),
    m_waveBankHandle(KWL_INVALID_HANDLE),
    m_currentBalance(0.0f),
    m_panBar(-1, 1, false)
{
    
}

BalanceAndPanDemo::~BalanceAndPanDemo()
{

}

void BalanceAndPanDemo::update(float timeStep)
{
    float delta = timeStep * 0.6f;
    Uint8* kstates = SDL_GetKeyState(NULL);
    
    if (kstates[SDLK_RIGHT])
    {
        m_currentBalance += delta;
    }
    else if (kstates[SDLK_LEFT])
    {
        m_currentBalance -= delta;
    }
    
    if (m_currentBalance > 1.0f)
    {
        m_currentBalance = 1.0f;
    }
    else if (m_currentBalance < -1.0f)
    {
        m_currentBalance = -1.0f;
    }
    
    kwlEventSetBalance(m_eventHandle, m_currentBalance);
}

void BalanceAndPanDemo::render2D()
{
    float x = KowalskiDemoManager::VIEWPORT_WIDTH / 2;
    float y = KowalskiDemoManager::VIEWPORT_HEIGHT / 2;
    float h = 30;
    float w = 400;
    
    m_panBar.setValue(m_currentBalance);
    m_panBar.render(x - w / 2, y - h / 2, w, h);
}

void BalanceAndPanDemo::initialize()
{
    m_currentBalance = 0.0f;
    m_waveBankHandle = kwlWaveBankLoad(getResourcePath("sfx.kwb"));
    m_eventHandle = kwlEventGetHandle("balancedemo/mono_and_stereo");
    kwlEventStartFade(m_eventHandle, 1.0f);
}

void BalanceAndPanDemo::deinitialize()
{
    kwlEventStop(m_eventHandle);
    kwlEventRelease(m_eventHandle);
    kwlWaveBankUnload(m_waveBankHandle);
    m_eventHandle = KWL_INVALID_HANDLE;
    m_waveBankHandle = KWL_INVALID_HANDLE;
}

void BalanceAndPanDemo::onKeyDown(SDLKey key)
{
    
}

const char* BalanceAndPanDemo::getInstructionLine(int index)
{
    switch (index)
    {
        case 0:
            return "The example event uses a sound with both mono and stereo data."; 
        case 1:
            return "Change pan/balance: left/right arrow keys";
    }
    return "";
}
